### ISSUE TITLE
Traceback "ENGINE Error in HTTPServer.tick" with unaccepted SSL/TLS connections

### ISSUE THREAD

**Issue by sanderjo** [sabnzbd/sabnzbd#820]
SABnzb 1.2.1Beta1 / Python 2.7.12 / Ubuntu 16.10

**TL;DR**: SABnzbd gives a **Traceback** in case of a problem with an incoming SSL/TLS connection. I would prefer an informing log line.

**Disclaimer**: I'm **assuming** it's OK that SAB refuses these connections, but I'm not sure

As soon as testssl.sh starts testing SABnbzd's HTTPS, SABnzbd spits out a lot (200!) of Tracebacks SSL error messages. See below.

`ENGINE Error in HTTPServer.tick Traceback (most recent call last):`
followed by
`SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:590)`
or
`SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:590)`

How relevant is this? This does happen too when normal HTTPS clients SABnzbd's HTTPS. See https://forums.sabnzbd.org/viewtopic.php?f=2&t=22369&p=110844 and maybe https://forums.sabnzbd.org/viewtopic.php?f=3&t=22425&p=110821 (different message)

I think the Tracebacks are too alarming. I think it would be better if there was just a informing message in sabnzbd.log (INFO or DEBUG)

```
2017-02-14 07:39:59,421::ERROR::[_cplogging:219] [14/Feb/2017:07:39:59] ENGINE Error in HTTPServer.tick
Traceback (most recent call last):
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/__init__.py", line 2024, in start
    self.tick()
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/__init__.py", line 2091, in tick
    s, ssl_env = self.ssl_adapter.wrap(s)
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/ssl_builtin.py", line 67, in wrap
    server_side=True)
  File "/usr/lib/python2.7/ssl.py", line 353, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 601, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 830, in do_handshake
    self._sslobj.do_handshake()
SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:590)

```
and

```
2017-02-14 07:39:59,475::ERROR::[_cplogging:219] [14/Feb/2017:07:39:59] ENGINE Error in HTTPServer.tick
Traceback (most recent call last):
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/__init__.py", line 2024, in start
    self.tick()
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/__init__.py", line 2091, in tick
    s, ssl_env = self.ssl_adapter.wrap(s)
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/ssl_builtin.py", line 67, in wrap
    server_side=True)
  File "/usr/lib/python2.7/ssl.py", line 353, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 601, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 830, in do_handshake
    self._sslobj.do_handshake()
SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:590)

```

**Comment by Safihre** [comment]
@sanderjo can you test the new branch:
`bugfix/ssl_catch`

**Comment by sanderjo** [comment]
The traceback on SAB's side is easy to reproduce:

Client side code:
```
import socket
import ssl

# SET VARIABLES
packet, reply = "<packet>SOME_DATA</packet>", ""
HOST, PORT = '127.0.0.1', 9090

# CREATE SOCKET
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.settimeout(10)

# WRAP SOCKET
wrappedSocket = ssl.wrap_socket(sock, ssl_version=ssl.PROTOCOL_TLSv1, ciphers="ADH-AES256-SHA")

# CONNECT AND PRINT REPLY
wrappedSocket.connect((HOST, PORT))
wrappedSocket.send(packet)
print wrappedSocket.recv(1280)

# CLOSE SOCKET CONNECTION
wrappedSocket.close()

```
Client reports:

```
$ python python_specific_cypher.py 
Traceback (most recent call last):
  File "python_specific_cypher.py", line 16, in <module>
    wrappedSocket.connect((HOST, PORT))
  File "/usr/lib/python2.7/ssl.py", line 866, in connect
    self._real_connect(addr, False)
  File "/usr/lib/python2.7/ssl.py", line 857, in _real_connect
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 830, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:590)

```
SAB's traceback:

```
2017-02-14 08:02:21,549::ERROR::[_cplogging:219] [14/Feb/2017:08:02:21] ENGINE Error in HTTPServer.tick
Traceback (most recent call last):
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/__init__.py", line 2024, in start
    self.tick()
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/__init__.py", line 2091, in tick
    s, ssl_env = self.ssl_adapter.wrap(s)
  File "/usr/share/sabnzbdplus/cherrypy/wsgiserver/ssl_builtin.py", line 67, in wrap
    server_side=True)
  File "/usr/lib/python2.7/ssl.py", line 353, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 601, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 830, in do_handshake
    self._sslobj.do_handshake()
SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:590)



```

**Comment by sanderjo** [comment]
FWIW: Tested with SABnzbd 1.1.1 and testssl.sh: server-side (=SABnzbd): 0 errors / messages.

(SABnzbd 1.2.0 gives the same error messages as SABnzbd 1.2.1Beta1)

BRB with the branch test.

FWIW: Statistics from sabnzbd.log after a few testssl.sh runs:

```
$ cat .sabnzbd/logs/sabnzbd.log | grep SSLError | sort | uniq -c
      4 SSLError: [SSL: CCS_RECEIVED_EARLY] ccs received early (_ssl.c:590)
      1 SSLError: [SSL: INAPPROPRIATE_FALLBACK] inappropriate fallback (_ssl.c:590)
    885 SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:590)
    110 SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:590)
```

**Comment by sanderjo** [comment]
Tested "remotes/origin/bugfix/ssl_catch" against testssl.sh: no traceback in SABnzbd, no errors in SABnzbd. 

Well done @Safihre 

So ... close this issue?

**Comment by sanderjo** [comment]
With the current git version of origin/1.2.x, while testing SAB with testssl.sh, SABnzbd's stdout says:

```
2017-02-21 20:52:28,914::ERROR::[_cplogging:219] [21/Feb/2017:20:52:28] ENGINE Error in HTTPServer.tick
Traceback (most recent call last):
  File "/home/sander/git/sab-19feb2017/cherrypy/wsgiserver/__init__.py", line 2024, in start
    self.tick()
  File "/home/sander/git/sab-19feb2017/cherrypy/wsgiserver/__init__.py", line 2091, in tick
    s, ssl_env = self.ssl_adapter.wrap(s)
  File "/home/sander/git/sab-19feb2017/cherrypy/wsgiserver/ssl_builtin.py", line 67, in wrap
    server_side=True)
  File "/usr/lib/python2.7/ssl.py", line 363, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 611, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 840, in do_handshake
    self._sslobj.do_handshake()
SSLError: [SSL] unknown error (_ssl.c:661)
```

"unknown error" ... ? I added that to cherrypy/wsgiserver/ssl_builtin.py:

```
                # Check if it's one of the known errors
                # Errors that are caught by PyOpenSSL, but thrown by built-in ssl
                _block_errors = ('unknown protocol', 'unknown ca', 'unknown_ca',
                                 'inappropriate fallback', 'wrong version number', 'unknown error',
                                 'no shared cipher', 'certificate unknown', 'ccs received early')
```
... and the SAB tracebacks are gone.

BUT ... is it OK to ignore 'unknown error'?

**Comment by Safihre** [comment]
Seems fine to me. Nothing we can do really if the connection can't be established.

**Comment by thezoggy** [comment]
@sanderjo @Safihre http://recollection.saaj.me/article/cherrypy-questions-testing-ssl-and-docker.html
> ssl_builtin doesn’t handle underlying ssl module exceptions correctly, as the log was flooded by exceptions originating mostly from do_handshake()

**Comment by thezoggy** [comment]
where is testssh.py ?
